File: plateform/frontend/src/api/map.ts
import type { AxiosRequestConfig } from "axios";

import { apiClient } from "./client";
import type {
  AssetCollection,
  IncidentCollection,
  DetectionCollection,
  MapFilters,
  MapSummary,
  NearbyResponse,
  PipelineCollection,
  ZoneCollection,
} from "../types/map";

export function toMapQuery(filters: MapFilters): Record<string, string> {
  const query: Record<string, string> = {};
  if (filters.zone) {
    query.zone = filters.zone;
  }
  if (filters.asset_type) {
    query.asset_type = filters.asset_type;
  }
  if (filters.asset_status) {
    query.asset_status = filters.asset_status;
  }
  if (filters.incident_severity) {
    query.incident_severity = filters.incident_severity;
  }
  if (filters.incident_status) {
    query.incident_status = filters.incident_status;
  }
  if (filters.include_resolved) {
    query.include_resolved = "true";
  }
  return query;
}

export async function fetchMapZones(
  filters: MapFilters,
  config?: AxiosRequestConfig,
): Promise<ZoneCollection> {
  const { data } = await apiClient.get<ZoneCollection>("/api/map/zones", {
    params: toMapQuery(filters),
    ...config,
  });
  return data;
}

export async function fetchMapPipelines(
  filters: MapFilters,
  config?: AxiosRequestConfig,
): Promise<PipelineCollection> {
  const { data } = await apiClient.get<PipelineCollection>("/api/map/pipelines", {
    params: toMapQuery(filters),
    ...config,
  });
  return data;
}

export async function fetchMapAssets(
  filters: MapFilters,
  config?: AxiosRequestConfig,
): Promise<AssetCollection> {
  const { data } = await apiClient.get<AssetCollection>("/api/map/assets", {
    params: toMapQuery(filters),
    ...config,
  });
  return data;
}

export async function fetchMapIncidents(
  filters: MapFilters,
  config?: AxiosRequestConfig,
): Promise<IncidentCollection> {
  const { data } = await apiClient.get<IncidentCollection>("/api/map/incidents", {
    params: toMapQuery(filters),
    ...config,
  });
  return data;
}

export async function fetchMapDetections(
  filters: MapFilters,
  config?: AxiosRequestConfig,
): Promise<DetectionCollection> {
  const { data } = await apiClient.get<DetectionCollection>("/api/map/detections", {
    params: toMapQuery(filters),
    ...config,
  });
  return data;
}

export async function fetchMapSummary(
  filters: MapFilters,
  config?: AxiosRequestConfig,
): Promise<MapSummary> {
  const { data } = await apiClient.get<MapSummary>("/api/map/summary", {
    params: toMapQuery(filters),
    ...config,
  });
  return data;
}

export async function fetchNearby(
  latitude: number,
  longitude: number,
  radius_m: number,
  featureType?: "asset" | "pipeline" | "incident",
  config?: AxiosRequestConfig,
): Promise<NearbyResponse> {
  const { data } = await apiClient.get<NearbyResponse>("/api/map/nearby", {
    params: {
      latitude,
      longitude,
      radius_m,
      ...(featureType ? { feature_type: featureType } : {}),
    },
    ...config,
  });
  return data;
}

File: plateform/backend/app/api/routes/map.py
from fastapi import APIRouter, Depends, Query

from app.api.deps import get_map_service
from app.schemas.assets import AssetType, OperationalStatus
from app.schemas.incidents import IncidentStatus, SeverityTier
from app.schemas.map import FeatureCollection, MapSummary, NearbyResponse
from app.services.map import MapService

router = APIRouter(prefix="/map", tags=["map"])

MAX_RADIUS_M = 25_000


@router.get("/zones", response_model=FeatureCollection)
def get_map_zones(
    zone: str | None = Query(default=None),
    service: MapService = Depends(get_map_service),
) -> FeatureCollection:
    return service.zones(zone=zone)


@router.get("/pipelines", response_model=FeatureCollection)
def get_map_pipelines(
    zone: str | None = Query(default=None),
    service: MapService = Depends(get_map_service),
) -> FeatureCollection:
    return service.pipelines(zone=zone)


@router.get("/assets", response_model=FeatureCollection)
def get_map_assets(
    zone: str | None = Query(default=None),
    asset_type: AssetType | None = Query(default=None),
    asset_status: OperationalStatus | None = Query(default=None),
    service: MapService = Depends(get_map_service),
) -> FeatureCollection:
    return service.assets(zone=zone, asset_type=asset_type, asset_status=asset_status)


@router.get("/incidents", response_model=FeatureCollection)
def get_map_incidents(
    zone: str | None = Query(default=None),
    incident_severity: SeverityTier | None = Query(default=None),
    incident_status: IncidentStatus | None = Query(default=None),
    include_resolved: bool = Query(default=False),
    service: MapService = Depends(get_map_service),
) -> FeatureCollection:
    return service.incidents(
        zone=zone,
        incident_severity=incident_severity,
        incident_status=incident_status,
        include_resolved=include_resolved,
    )


@router.get("/detections", response_model=FeatureCollection)
def get_map_detections(
    zone: str | None = Query(default=None),
    service: MapService = Depends(get_map_service),
) -> FeatureCollection:
    return service.detections(zone=zone)


@router.get("/summary", response_model=MapSummary)
def get_map_summary(
    zone: str | None = Query(default=None),
    asset_type: AssetType | None = Query(default=None),
    asset_status: OperationalStatus | None = Query(default=None),
    incident_severity: SeverityTier | None = Query(default=None),
    incident_status: IncidentStatus | None = Query(default=None),
    include_resolved: bool = Query(default=False),
    service: MapService = Depends(get_map_service),
) -> MapSummary:
    return service.summary(
        zone=zone,
        asset_type=asset_type,
        asset_status=asset_status,
        incident_severity=incident_severity,
        incident_status=incident_status,
        include_resolved=include_resolved,
    )


@router.get("/nearby", response_model=NearbyResponse)
def get_map_nearby(
    latitude: float = Query(..., ge=-90, le=90),
    longitude: float = Query(..., ge=-180, le=180),
    radius_m: float = Query(500, gt=0, le=MAX_RADIUS_M),
    feature_type: str | None = Query(default=None, pattern="^(asset|pipeline|incident)$"),
    service: MapService = Depends(get_map_service),
) -> NearbyResponse:
    return service.nearby(
        latitude=latitude,
        longitude=longitude,
        radius_m=radius_m,
        feature_type=feature_type,
    )
